[ASCollectionView / ASTableView] Optimize reloadData and reloadSection: methods.

lkzhao opened the pull request:
Calls reloadData & reloadSection on UITableView/UICollectionView instead of deleting/inserting one by one.  
Similar to #1093.

Please review this but do not merge yet. still seeing that bug where collectionView will cause inconsistent crash when inserting rows unless [collectionView numberOfSections] is called before inserting.
@nguyenhuy 
@appleguy

appleguy commented:
@lkzhao this is great, very glad to see this! Are there any areas in particular that I should focus on for review?  @nguyenhuy 's original disk to attempt something similar was reverted a couple times, but I can't remember if we ever really landed it after the second time… this this is a attempting to do something similar with a new approach, correct? Just want to make sure I understand where the prior one left off, it will not re-land again?

lkzhao commented:
@appleguy Yes, both are optimizing reload calls. This patch includes the other one that Huy made and fixed the crashes we had earlier. I actually took some code directly from his fork. We can land this instead. 


Few notes about what I did here:

1. Removed `_deleteSection`, `_insertSection` etc. These don't do much and made it very confusing to read the code.

2. **(Please review this part)** Every node editing methods previously all have a completion callback scheduled on the main thread and update the _completedNodes on the main thread. This made it hard and cause crashes when doing batch operations like ReloadSection/ReloadData etc. 
Now these methods only change _editingNodes and do nothing with the _completedNodes. After we are done with the operations, we call `commitChangesToNodesOfKind ` and provide a callback there. 

3. Optimized reloadSection/reloadData/reloadItems/deleteSection/insertSections on TableView and CollectionView. Previously, all of these actions result in deleteItems/insertItems calls to TableView and CollectionView. Now it makes one specific call. (for example, reloadSection calls tableview `reloadSections: withRowAnimation` instead of making `deleteRowsAtIndexPaths: and insertRowsAtIndexPaths:`)

appleguy commented:
Thank you Luke, this summary was extremely helpful for me.  I will review in detail soon.

Sent from my iPhone

> On Feb 6, 2016, at 10:06 AM, Luke Zhao <notifications@github.com> wrote:
> 
> @appleguy Yes, both are optimizing reload calls. This patch includes the other one that Huy made and fixed the crashes we had earlier. I actually took some code directly from his fork. We can land this instead.
> 
> Few notes about what I did here:
> 
> Removed _deleteSection, _insertSection etc. These don't do much and made it very confusing to read the code.
> 
> (Please review this part) Every node editing methods previously all have a completion callback scheduled on the main thread and update the _completedNodes on the main thread. This made it hard and cause crashes when doing batch operations like ReloadSection/ReloadData etc. 
> Now these methods only change _editingNodes and do nothing with the _completedNodes. After we are done with the operations, we call commitChangesToNodesOfKind and provide a callback there.
> 
> Optimized reloadSection/reloadData/reloadItems/deleteSection/insertSections on TableView and CollectionView. Previously, all of these actions result in deleteItems/insertItems calls to TableView and CollectionView. Now it makes one specific call. (for example, reloadSection calls tableview reloadSections: withRowAnimation instead of making deleteRowsAtIndexPaths: and insertRowsAtIndexPaths:)
> 
> —
> Reply to this email directly or view it on GitHub.
>

nguyenhuy commented on AsyncDisplayKit/ASCollectionView.mm:
```diff
@@ -962,6 +1002,23 @@ - (void)rangeController:(ASRangeController *)rangeController didDeleteSectionsAt
+  }
+
+  if (_performingBatchUpdates) {
+    [_batchUpdateBlocks addObject:^{
```
Do we need to notify the `_layoutFacilitator` here? cc @binl

binl replied:
yes, we will need to do
```
[_layoutFacilitator collectionViewWillEditCellsAtIndexPaths:@[[NSIndexPath indexPathForItem:0 section:0]] batched:YES/NO];
```

nguyenhuy commented on AsyncDisplayKit/ASTableView.mm:
```diff
@@ -303,7 +303,7 @@ - (void)reloadDataWithCompletion:(void (^)())completion
   ASPerformBlockOnMainThread(^{
     [super reloadData];
```
Not sure if possible, but I really want to eliminate this call. We will do this afterward anyway.

nguyenhuy commented on AsyncDisplayKit/ASTableView.mm:
```diff
@@ -314,7 +314,7 @@ - (void)reloadData
   ASDisplayNodeAssertMainThread();
-  [_dataController reloadDataImmediatelyWithAnimationOptions:UITableViewRowAnimationNone];
+  [_dataController reloadDataImmediately];
   [super reloadData];
```
Same as above

nguyenhuy commented on AsyncDisplayKit/ASTableView.mm:
```diff
@@ -831,6 +831,25 @@ - (void)rangeController:(ASRangeController *)rangeController didDeleteNodes:(NSA
+    [super reloadRowsAtIndexPaths:indexPaths withRowAnimation:(UITableViewRowAnimation)animationOptions];
+  });
+
+  if (_automaticallyAdjustsContentOffset) {
```
Not sure what is the policy to adjust content offset? It is done in after reloading nodes, but not sections or everything/data.

appleguy replied:
@nguyenhuy @lkzhao I think we should honor it wherever it can be reasonably accommodated, especially inserting and deleting elements, as well as moving or reloading individual items and sections as that may cause them to calculate a different height. I think the main challenge with doing it on a full table reload at the moment is that we break it into a delete and insert, At least historically, and because of the asynchronous delay of the insert it was not always feasible to guarantee.

nguyenhuy replied:
Now it is feasible thanks to this awesome diff. Let's do it then!

nguyenhuy commented on AsyncDisplayKit/Details/ASCollectionDataController.mm:
```diff
@@ -58,11 +58,11 @@ - (void)willReloadData
     NSArray *editingNodes = [self editingNodesOfKind:kind];
-    NSMutableIndexSet *indexSet = [[NSMutableIndexSet alloc] initWithIndexesInRange:NSMakeRange(0, editingNodes.count)];
-    [self deleteSectionsOfKind:kind atIndexSet:indexSet completion:nil];
+    NSMutableIndexSet *indexSet = [[NSMutableIndexSet alloc]initWithIndexesInRange:NSMakeRange(0, editingNodes.count)];
```
nit: white space before `initWithIndexesInRange`

nguyenhuy commented on AsyncDisplayKit/Details/ASDataController.h:
```diff
@@ -85,15 +85,30 @@ FOUNDATION_EXPORT NSString * const ASDataControllerRowNodeKind;
+/**
+ Called for reload of sections.
+ */
+- (void)dataController:(ASDataController *)dataController didReloadSectionsAtIndexSet:(NSIndexSet *)indexSet withAnimationOptions:(ASDataControllerAnimationOptions)animationOptions;
```
Maybe just `didReloadSections:(NSIndexSet *)sections`?

lkzhao replied:
Tableview reloadSection actually supports animation. I think we should keep the animationOption here.

nguyenhuy replied:
I meant the naming of that particular sections param.

lkzhao replied:
The ASRangeController and ASDataController all uses `didXXXXSectionsAtIndexSet` pattern. I think we should keep it consistent.

for example:
```
- (void)rangeController:(ASRangeController *)rangeController didDeleteSectionsAtIndexSet:(NSIndexSet *)indexSet withAnimationOptions:
```

nguyenhuy commented on AsyncDisplayKit/Details/ASDataController.h:
```diff
@@ -85,15 +85,30 @@ FOUNDATION_EXPORT NSString * const ASDataControllerRowNodeKind;
  Called for insertion of sections.
  */
-- (void)dataController:(ASDataController *)dataController didInsertSections:(NSArray<NSArray<ASCellNode *> *> *)sections atIndexSet:(NSIndexSet *)indexSet withAnimationOptions:(ASDataControllerAnimationOptions)animationOptions;
+- (void)dataController:(ASDataController *)dataController didInsertSectionsAtIndexSet:(NSIndexSet *)indexSet withAnimationOptions:(ASDataControllerAnimationOptions)animationOptions;
```
Same as below, `didInsertSections:(NSIndexSet)sections`?

nguyenhuy commented on AsyncDisplayKit/Details/ASDataController.mm:
```diff
@@ -118,7 +127,7 @@ - (void)batchLayoutNodes:(NSArray *)nodes ofKind:(NSString *)kind atIndexPaths:(
     NSArray *batchedNodes = [nodes subarrayWithRange:batchedRange];
     
-    [self _layoutNodes:batchedNodes ofKind:kind atIndexPaths:batchedIndexPaths completion:completionBlock];
+    [self _layoutNodes:batchedNodes ofKind:kind atIndexPaths:batchedIndexPaths ];
```
nit: no whitespace after `atIndexPaths:batchedIndexPaths`

nguyenhuy commented on AsyncDisplayKit/Details/ASRangeControllerBeta.mm:
```diff
@@ -275,15 +283,22 @@ - (void)dataController:(ASDataController *)dataController didDeleteNodes:(NSArra
-- (void)dataController:(ASDataController *)dataController didInsertSections:(NSArray *)sections atIndexSet:(NSIndexSet *)indexSet withAnimationOptions:(ASDataControllerAnimationOptions)animationOptions
+- (void)dataController:(ASDataController *)dataController didInsertSectionsAtIndexSet:(NSIndexSet *)indexSet withAnimationOptions:(ASDataControllerAnimationOptions)animationOptions
 {
-  ASDisplayNodeAssert(sections.count == indexSet.count, @"Invalid sections");
```
Shouldn't we keep this assert?

lkzhao replied:
I eleminated the section parameter on this call. therefore no need to check length equality anymore.

nguyenhuy replied:
oh, right. I missed that :)

nguyenhuy commented on AsyncDisplayKit/Details/ASDataController.mm:
```diff
@@ -391,30 +317,26 @@ - (void)_reloadDataWithAnimationOptions:(ASDataControllerAnimationOptions)animat
-        
-        if (completion) {
-          dispatch_async(dispatch_get_main_queue(), completion);
+
```
Should probably call `[self willReloadData];` here.

lkzhao replied:
Nice catch

nguyenhuy commented on AsyncDisplayKit/Details/ASDataController.mm:
```diff
@@ -391,30 +317,26 @@ - (void)_reloadDataWithAnimationOptions:(ASDataControllerAnimationOptions)animat
+
+        [self batchLayoutNodes:updatedNodes ofKind:ASDataControllerRowNodeKind atIndexPaths:updatedIndexPaths];
+
+        [self commitChangesToNodesOfKind:ASDataControllerRowNodeKind withCompletion:^{
```
Small optimization here: can pass `NIL` here if `_delegateDidReloadData` is `NO` and `completion` is `nil`. That way we don't schedule a block on main thread that essentially does nothing.

nguyenhuy commented:
I have 2 main concerns/feedbacks about this PR:

1. We should clearly document a few things for *each* of the methods in `ASDataController+Subclasses.h`:
  - It should (and is assuming to) be called on `_editingTransactionQueue`.
  - It blocks the calling (background) thread.
  - It only modifies `_editingNodes`.
  - Its caller is expected to commit the changes to `_completedNodes` on main afterward.

2. I think adding `commitChangesToNodesOfKind` is a really good idea that helps to simplify the controller. However, deep copying a 2D array is expensive and should be avoided. That's why in the current implementation, we don't deep copy after deletes, but instead directly remove items. Can we do the same thing here? 

Other than that, I reckon this new implementation should be faster than the one on master because we update `_completedNodes` and notify delegate once after all operations finished :+1:

lkzhao commented:
@nguyenhuy Thanks for the review & feedbacks!! There are some code that I copied/converted and I am not entirely sure what they do. Let me talk to you tomorrow about some of the details int your review.

appleguy commented:
Very excited to land this!  Thank you @lkzhao
